File: src/InfiniFrame/Window/InfiniFrameWindowFeatures.cs
// ---------------------------------------------------------------------------------------------------------------------
// Imports
// ---------------------------------------------------------------------------------------------------------------------
namespace InfiniFrame;
// ---------------------------------------------------------------------------------------------------------------------
// Code
// ---------------------------------------------------------------------------------------------------------------------
// ReSharper disable InvalidXmlDocComment

/// <summary>
///     Aggregates all feature instances available for an <see cref="IInfiniFrameWindow" />.
/// </summary>
public sealed record InfiniFrameWindowFeatures(
    /// <inheritdoc cref="IInfiniFrameWindowFeatures.Debugging"/>
    IDebuggingInfiniFrameWindowFeature Debugging,
    /// <inheritdoc cref="IInfiniFrameWindowFeatures.Lifecycle"/>
    ILifecycleInfiniFrameWindowFeature Lifecycle,
    /// <inheritdoc cref="IInfiniFrameWindowFeatures.Invoke"/>
    IInvokeInfiniFrameWindowFeature Invoke,
    /// <inheritdoc cref="IInfiniFrameWindowFeatures.WebMessaging"/>
    IWebMessagingInfiniFrameWindowFeature WebMessaging,
    /// <inheritdoc cref="IInfiniFrameWindowFeatures.Notifications"/>
    INotificationsInfiniFrameWindowFeature Notifications,
    /// <inheritdoc cref="IInfiniFrameWindowFeatures.FilePickerDialogs"/>
    IFilePickerDialogsInfiniFrameWindowFeature FilePickerDialogs,
    /// <inheritdoc cref="IInfiniFrameWindowFeatures.Monitors"/>
    IMonitorsInfiniFrameWindowFeature Monitors,
    /// <inheritdoc cref="IInfiniFrameWindowFeatures.PageNavigation"/>
    IPageNavigationInfiniFrameWindowFeature PageNavigation,
    /// <inheritdoc cref="IInfiniFrameWindowFeatures.Position"/>
    IPositionInfiniFrameWindowFeature Position,
    /// <inheritdoc cref="IInfiniFrameWindowFeatures.Size"/>
    ISizeInfiniFrameWindowFeature Size,
    /// <inheritdoc cref="IInfiniFrameWindowFeatures.Decorations"/>
    IDecorationsInfiniFrameWindowFeature Decorations,
    /// <inheritdoc cref="IInfiniFrameWindowFeatures.State"/>
    IStateInfiniFrameWindowFeature State,
    /// <inheritdoc cref="IInfiniFrameWindowFeatures.Browser"/>
    IBrowserInfiniFrameWindowFeature Browser
) : IInfiniFrameWindowFeatures;

File: src/InfiniFrame.Js/TypeScript/Contracts/Window/InfiniFrameWindowFeatures.ts
/**
 * Window features collection contract. Defines the shape of window.infiniframe.window.features.
 * @module InfiniFrameWindowFeatures
 */

// ---------------------------------------------------------------------------------------------------------------------
// Imports
// ---------------------------------------------------------------------------------------------------------------------
import {
    BrowserInfiniFrameWindowFeature,
    DebuggingInfiniFrameWindowFeature,
    DecorationsInfiniFrameWindowFeature,
    FilePickerDialogsInfiniFrameWindowFeature,
    InvokeInfiniFrameWindowFeature,
    LifecycleInfiniFrameWindowFeature,
    MonitorsInfiniFrameWindowFeature,
    NotificationsInfiniFrameWindowFeature,
    PageNavigationInfiniFrameWindowFeature,
    PositionInfiniFrameWindowFeature,
    SizeInfiniFrameWindowFeature,
    StateInfiniFrameWindowFeature,
    WebMessagingInfiniFrameWindowFeature
} from "./Features";

// ---------------------------------------------------------------------------------------------------------------------
// Code
// ---------------------------------------------------------------------------------------------------------------------

/**
 * Aggregates all window feature groups available on an InfiniFrame window instance.
 * Each property provides a focused API for a specific window capability.
 */
export interface InfiniFrameWindowFeatures {
    /** Browser-level settings: context menu, media autoplay, user agent, web security, etc. */
    browser: BrowserInfiniFrameWindowFeature
    /** Debugging tools: dev tools, remote debugging endpoint, web inspector, diagnostics. */
    debugging: DebuggingInfiniFrameWindowFeature
    /** Window decorations: title, icon, background color, transparency, chromeless mode. */
    decorations: DecorationsInfiniFrameWindowFeature
    /** Native file picker dialogs: open file, open folder, save file. */
    filePickerDialogs: FilePickerDialogsInfiniFrameWindowFeature
    /** Direct method invocation on the native host (reserved for future use). */
    invoke: InvokeInfiniFrameWindowFeature
    /** Window lifecycle management: state queries, close, and lifecycle events. */
    lifecycle: LifecycleInfiniFrameWindowFeature
    /** Multi-monitor support: enumerate monitors, get DPI, query work areas. */
    monitors: MonitorsInfiniFrameWindowFeature
    /** Notifications and dialog boxes: toast notifications, message dialogs. */
    notifications: NotificationsInfiniFrameWindowFeature
    /** Page navigation: load URIs, load paths, load raw HTML, get current URL. */
    pageNavigation: PageNavigationInfiniFrameWindowFeature
    /** Window position: get/set location, center, offset, and monitor-aware placement. */
    position: PositionInfiniFrameWindowFeature
    /** Window size: get/set dimensions, min/max constraints, resize, and resizable flag. */
    size: SizeInfiniFrameWindowFeature
    /** Window state: fullscreen, maximize, minimize, topmost, focus, zoom. */
    state: StateInfiniFrameWindowFeature
    /** Web messaging: send web messages to the native host. */
    webMessaging: WebMessagingInfiniFrameWindowFeature
}
